File: python/016_three_sum_closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        temp = 1e100
        for i in range(len(nums)):
            for j in range(i+1, len(nums)):
                for k in range(j+1, len(nums)):
                    if abs(nums[i] + nums[j] + nums[k] - target) < temp:
                        temp = abs(nums[i] + nums[j] + nums[k] - target)
                        score = nums[i] + nums[j] + nums[k]
        return score


class Solution2(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        score = sum(nums[:3])
        if len(nums) == 3:
            return score

        for i in range(len(nums)-1):
            j, k = i+1, len(nums)-1
            while j < k:
                temp = nums[i] + nums[j] + nums[k]
                if abs(temp - target) < abs(score - target):
                    score = temp
                if temp <= target:
                    j += 1
                elif temp > target:
                    k -= 1
        return score


if __name__ == '__main__':
    s = Solution()
    print(s.threeSumClosest([-1, 2, 1, -4], 1))

File: cpp/016_three_sum_closest.cpp
#include "header.h"
using namespace  std;

// two pointer
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        int ans = nums[0] + nums[1] + nums[2];
        if (ans>=target) return ans;
        for (int i=0; i<n-2; ++i) {
            int l = i+1;
            int r = n-1;
            while (l<r) {
                if (abs(target-ans)>abs(target-nums[i]-nums[l]-nums[r])) ans = nums[i]+nums[l]+nums[r];
                if (nums[i]+nums[l]+nums[r]>target) --r; 
                else if (nums[i]+nums[l]+nums[r]<target) ++l;
                else return target;
            }
        }
        return ans;
    }
};

int main() {
    Solution s;
    vector<int> nums = {-1, 2, 1, -4};
    cout << s.threeSumClosest(nums, 1) << endl;
}